From 24137307b60446659d41ab29b31404a389b35cda Mon Sep 17 00:00:00 2001
From: Varun Kumar Reddy Yaparla <vyaparla@habana.ai>
Date: Fri, 17 Dec 2021 07:09:09 +0200
Subject: [PATCH] [SW-67082] gaudi2: fake security support for signature validation

This change adds fake security support for image signature
validation check. In case of real security, signature validation
bit is read from efuse and in case of fake security, it is read
from fake security boot info

Change-Id: I0bfd7a4f4a180728fc58512ddd7f789b6118af5a
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 13ae454..85f4808 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -474,5 +474,8 @@
 
 bool preboot_img_validation_is_enabled(void)
 {
-	return preboot_sec_efuse_is_set(EFUSE_SYSCFG_ENABLE_VALIDATION_MASK);
+	if (preboot_sec_efuse_is_set(EFUSE_SYSCFG_ENABLE_VALIDATION_MASK))
+		return true;
+
+	return preboot_fake_security_is_sig_validation_enabled();
 }
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 192103c..32e04ea 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -444,5 +444,8 @@
 
 bool preboot_img_validation_is_enabled(void)
 {
-	return preboot_sec_efuse_is_set(EFUSE_SECURITY_IMG_VALIDATION_BIT);
+	if (preboot_sec_efuse_is_set(EFUSE_SECURITY_IMG_VALIDATION_BIT))
+		return true;
+
+	return preboot_fake_security_is_sig_validation_enabled();
 }
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index d863c59..39906e7 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -31,6 +31,7 @@
 #include <cmd_module.h>
 #include <gaudi2/gaudi2_reg_map.h>
 #include <gaudi2/gaudi2_boot_info.h>
+#include <gaudi2/gaudi2_security.h>
 #include <psoc_reset_conf_regs.h>
 #include <psoc_reset_conf_masks.h>
 #include <gaudi2_blocks_linux_driver.h>
@@ -134,6 +135,19 @@
 	return cold_rst_data.fake_security_enable;
 }
 
+bool preboot_fake_security_is_sig_validation_enabled(void)
+{
+	u32 fake_security;
+	u32 copy_sz;
+	bool sig_validation_en = false;
+
+	if (!boot_info_get(HL_BOOT_INFO_FAKE_SECURITY, &fake_security, sizeof(fake_security),
+				&copy_sz))
+		sig_validation_en = !!(fake_security & FAKE_SECURITY_SIG_VALIDATION_ENABLE);
+
+	return sig_validation_en;
+}
+
 void lkd_dev_spec_regs(struct cpu_dyn_regs *regs)
 {
 	regs->cpu_pq_base_addr_low = mmCPU_IF_PQ_BASE_ADDR_LOW;
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index 35e2c12..04cfe2e 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -197,4 +197,5 @@
 
 void gaudi2_axi_drain_ena(void);
 bool preboot_fake_security_is_enabled(void);
+bool preboot_fake_security_is_sig_validation_enabled(void);
 #endif /* ARCH_GAUDI2_COMMON_COMMON_H_ */
